### PR TITLE
Reshape

### PR DESCRIPTION
**Pull request by jwittenbach**
Partially addresses the second half of #7.

Right now, only reshapes that can be implemented by independent reshapes on the keys and/or values are allowed.

### CODE REVIEW

File: bolt/spark/array.py [hunk 1]
```diff
@@ -467,6 +469,42 @@ def swapaxes(self, ax1, ax2):
...
+        Check if the requested reshape can be broken into independant reshapes on the keys and values.
+        If it can, returns the index in the new shape separating keys from values.
+        If it cannot, returns -1
+        """
+
+        new = tupleize(new)
+        old_key_size = prod(self.keys.shape)
+        old_value_size = prod(self.values.shape)
+
+        for i in xrange(len(new)):
```
Comment (freeman-lab) [inline bolt/spark/array.py]: Change `xrange` -> `range`

File: bolt/utils.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
-from numpy import ndarray, asarray, any
+from numpy import ndarray, asarray, prod
+from numpy import any as np_any
```
Comment (freeman-lab) [inline bolt/utils.py]: let's use `npany` instead of `np_any`

File: test/spark/test_spark_shaping.py [hunk 3]
```diff
@@ -217,3 +217,22 @@ def test_swapaxes(sc):
     assert allclose(b.swapaxes(0, 1).toarray(), b.toarray().swapaxes(0, 1))
     assert allclose(b.swapaxes(2, 3).toarray(), b.toarray().swapaxes(2, 3))
 
+def test_reshape(sc):
+
+    old_shape = (6, 10, 4, 12)
+    a = arange(prod(old_shape)).reshape(old_shape)
+    b = array(a, sc, axis=[0, 1])
```
Comment (freeman-lab) [inline test/spark/test_spark_shaping.py]: swap `(0,1)` for `[0,1]`, I know it's the same, but we've used the tuple version throughout the tests

File: bolt/utils.py [hunk 4]
```diff
@@ -127,3 +128,38 @@ def slicify(slc, dim):
     else:
         raise ValueError("Type for slice %s not recongized" % type(slc))
 
+def prime_factors(n):
```
Comment (freeman-lab) [inline bolt/utils.py]: I'd hold off adding this for now, as it's not used in the codepath introduced in this PR. We'll hopefully bring it back soon!

File: bolt/spark/array.py [hunk 5]
```diff
@@ -467,6 +469,42 @@ def swapaxes(self, ax1, ax2):
...
+            return self
+
+        i = self._simple_reshapeable(new)
+        if i == -1:
+            raise NotImplementedError("Currently no support for reshaping between keys and values for BoltArraySpark")
+        else:
+            new_key_shape, new_value_shape = new[:i], new[i:]
+            return self.keys.reshape(new_key_shape).values.reshape(new_value_shape)
+
+    def _simple_reshapeable(self, new):
```
Comment (freeman-lab) [inline bolt/spark/array.py]: maybe rename to `_reshapebasic`? that would make it analogous to `_getbasic` and `_getadvanced`, which similarly breaks the core operation into two submethods depending on the complexity

**Comment by freeman-lab** [comment]
@jwittenbach this looks close! I like the strategy of implementing the simple case initially. I left a bunch of minor comments, and it'd be great to merge in the latest changes from master, which will run the tests on Python 3.4 as well, so we can identify any potential issues there.

**Comment by jwittenbach** [comment]
All of the comments make sense -- changes made!

File: bolt/spark/array.py [hunk 6]
```diff
@@ -568,6 +570,42 @@ def swapaxes(self, ax1, ax2):
 
         return self.transpose(p)
 
+    def reshape(self, *new):
```
Comment (freeman-lab) [inline bolt/spark/array.py]: change `new` to `shape`, matches numpy

Comment (jwittenbach) [inline bolt/spark/array.py]: I took this from your `transpose` function on the `Shapes` objects. Should we change it there too then?

Comment (freeman-lab) [inline bolt/spark/array.py]: In `transpose` it's called `axes`, no?

Comment (freeman-lab) [inline bolt/spark/array.py]: when in doubt, numpy is right and I'm wrong =)

**Comment by freeman-lab** [comment]
Awesome, great job tracking down those glitches. LGTM! Merging in.

**Pull request merged**